src/main/java/net/greghaines/jesque/worker/Watchdog.java

```diff
@@ -41,28 +48,39 @@
             Executors.newScheduledThreadPool(POOL_SIZE);
     private final AtomicReference<String> watchdogScriptHash = new AtomicReference<>(null);
     private Pool<Jedis> jedisPool;
+    private boolean redisRestartRecoveryOn;
+    private Runnable listener;
+    private String serverName;
+    private String isDebug;
+    private Future<?> recovering = null;
 
     private Watchdog() {
     }
 
     public static synchronized void activate(Config config) {
+        activate(config, null);
+    }
+
+    public static synchronized void activate(Config config, Runnable listener) {
         if (isActivated) {
             throw new RuntimeException("Watchdog was already activated with config " + config);
         }
 
-        instance.init(config);
+        instance.init(config, listener);
         isActivated = true;
     }
 
-    private void init(Config config) {
+    private void init(Config config, Runnable listener) {
         jedisPool = PoolUtils.createJedisPool(config, new WatchdogJedisPoolConfig());
-        final String isDebug = LOG.isDebugEnabled() ? Boolean.TRUE.toString() : Boolean.FALSE.toString();
-        final String serverName = getServerName();
+        isDebug = LOG.isDebugEnabled() ? Boolean.TRUE.toString() : Boolean.FALSE.toString();
+        serverName = getServerName();
+        this.listener = listener;
+        redisRestartRecoveryOn = (listener != null);
```

Comment:
Looks like we can avoid this and look at the listener instead.

Reply:
Here I went for best readability (IMO, of course).  redisRestartRecoveryOn is calculated one time , but used many time. I don't like checking whether some variable is null, I prefer it to be done only on initialization. I would like to use Optional instead, but optional is recommended only for method return value